sw/device/lib/crypto/drivers/entropy.c

```diff
@@ -0,0 +1,183 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/drivers/entropy.h"
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+
+#include "csrng_regs.h"  // Generated
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+enum {
+  kBaseCsrng = TOP_EARLGREY_CSRNG_BASE_ADDR,
+
+  /**
+   * CSRNG genbits buffer size in uint32_t words.
+   */
+  kEntropyCsrngBitsBufferSize = 4,
+};
+
+/**
+ * Supported CSRNG application commands.
+ * See https://docs.opentitan.org/hw/ip/csrng/doc/#command-header for
+ * details.
+ */
+typedef enum entropy_csrng_op {
+  kEntropyDrbgOpInstantiate = 1,
+  kEntropyDrbgOpReseed = 2,
+  kEntropyDrbgOpGenerate = 3,
+  kEntropyDrbgOpUpdate = 4,
+  kEntropyDrbgOpUnisntantiate = 5,
+} entropy_csrng_op_t;
```

Comment:
should these be hardened in hardware? can we try something like what @msfschaffner did for life cycle controller state values?

Reply:
This is now being tracked here: https://github.com/lowRISC/opentitan/issues/14542. I added a cross-reference to the issue in a TODO.